### PR TITLE
TEZ-4062. Speculative attempt scheduling should be aborted when Task has completed.

### PR DESCRIPTION
**Pull request by Chyler**

### CODE REVIEW

File: tez-dag/src/test/java/org/apache/tez/dag/app/dag/impl/TestTaskImpl.java [hunk 1]
```diff
@@ -982,6 +982,23 @@ public void testSpeculatedThenRetroactiveFailure() {
     Assert.assertEquals(mockDestId, newAttempt.getSchedulingCausalTA());
   }
 
+  @Test(timeout = 20000)
+  public void testIgnoreSpeculation() {
```
Comment (yingdachen) [inline tez-dag/src/test/java/org/apache/tez/dag/app/dag/impl/TestTaskImpl.java]: testIgnoreSpeculationOnSuccessfulOriginalAttempt

File: tez-dag/src/test/java/org/apache/tez/dag/app/dag/impl/TestTaskImpl.java [hunk 2]
```diff
@@ -982,6 +982,23 @@ public void testSpeculatedThenRetroactiveFailure() {
...
+    TezTaskID taskId = getNewTaskID();
+    scheduleTaskAttempt(taskId);
+    MockTaskAttemptImpl firstAttempt = mockTask.getLastAttempt();
+    launchTaskAttempt(firstAttempt.getID());
+    // Have the first task succeed
```
Comment (yingdachen) [inline tez-dag/src/test/java/org/apache/tez/dag/app/dag/impl/TestTaskImpl.java]: mock success of the first task attempt

File: tez-dag/src/test/java/org/apache/tez/dag/app/dag/impl/TestTaskImpl.java [hunk 3]
```diff
@@ -982,6 +982,23 @@ public void testSpeculatedThenRetroactiveFailure() {
...
+    updateAttemptState(firstAttempt, TaskAttemptState.SUCCEEDED);
+    firstAttempt.handle(new TaskAttemptEvent(firstAttempt.getID(), TaskAttemptEventType.TA_DONE));
+
+    // Ignore speculation
```
Comment (yingdachen) [inline tez-dag/src/test/java/org/apache/tez/dag/app/dag/impl/TestTaskImpl.java]: // verify the speculation scheduding is ignored and no speculative attempt was added to the task

File: tez-dag/src/test/java/org/apache/tez/dag/app/dag/impl/TestTaskImpl.java (hunk 1 above)
Comment (Chyler) [reply tez-dag/src/test/java/org/apache/tez/dag/app/dag/impl/TestTaskImpl.java]: done

File: tez-dag/src/test/java/org/apache/tez/dag/app/dag/impl/TestTaskImpl.java (hunk 2 above)
Comment (Chyler) [reply tez-dag/src/test/java/org/apache/tez/dag/app/dag/impl/TestTaskImpl.java]: done

File: tez-dag/src/test/java/org/apache/tez/dag/app/dag/impl/TestTaskImpl.java (hunk 3 above)
Comment (Chyler) [reply tez-dag/src/test/java/org/apache/tez/dag/app/dag/impl/TestTaskImpl.java]: done

**Pull request closed without merging**